### PR TITLE
Add configuration options to determine handling of missing components (#52)

### PR DESCRIPTION
**Pull request by thibaudcolas**
See #52

### CODE REVIEW

File: draftjs_exporter/composite_decorators.py [hunk 1]
```diff
@@ -31,16 +31,20 @@ def apply_decorators(decorators, text, block_type):
 
         yield DOM.create_element(decorator, {
             'match': match,
-            'block_type': block_type,
+            'block': {
+                'type': block['type'],
+                'depth': block['depth'],
+                'data': block.get('data', {}),
+            }
```
Comment (thibaudcolas) [inline draftjs_exporter/composite_decorators.py]: Aligning with the data structure of blocks.

File: draftjs_exporter/defaults.py [hunk 2]
```diff
@@ -3,6 +3,19 @@
...
+def render_children(props):
+    """
+    Renders the children of a component without any specific
+    markup for the component itself.
+    """
+    return props['children']
+
+
+def CodeBlock(props):
+    return DOM.create_element('pre', {}, DOM.create_element('code', {}, props['children']))
```
Comment (thibaudcolas) [inline draftjs_exporter/defaults.py]: Unrelated changes.

File: draftjs_exporter/entity_state.py [hunk 3]
```diff
@@ -39,28 +40,22 @@ def get_entity_details(self, entity_key):
...
-        return decorator
-
-    def render_entitities(self, style_node):
+    def render_entities(self, style_node):
 
         if self.completed_entity is not None:
-            # self.element_stack.append(style_node)
             entity_details = self.get_entity_details(self.completed_entity)
-            decorator = self.get_entity_decorator(entity_details['type'])
+            opts = Options.for_entity(self.entity_decorators, entity_details['type'])
```
Comment (thibaudcolas) [inline draftjs_exporter/entity_state.py]: Wasn't strictly required, but I thought it would be nicer to align the config handling of all types of "components".

File: draftjs_exporter/html.py [hunk 4]
```diff
@@ -64,12 +64,12 @@ def render_block(self, block, entity_map, wrapper_state):
...
             # Decorators are not rendered inside entities.
             if entity_state.has_no_entity() and len(self.composite_decorators) > 0:
-                decorated_node = render_decorators(self.composite_decorators, text, block['type'])
+                decorated_node = render_decorators(self.composite_decorators, text, block)
             else:
                 decorated_node = text
 
             styled_node = style_state.render_styles(decorated_node)
-            entity_node = entity_state.render_entitities(styled_node)
+            entity_node = entity_state.render_entities(styled_node)
```
Comment (thibaudcolas) [inline draftjs_exporter/html.py]: Silly typo.

File: draftjs_exporter/options.py [hunk 5]
```diff
@@ -30,37 +32,38 @@ def __ne__(self, other):
         return not self == other
 
     @staticmethod
-    def for_block(block_map, type_):
-        if type_ not in block_map:
-            raise ConfigException('Block "%s" does not exist in block_map' % type_)
-
-        block = block_map.get(type_)
+    def for_kind(kind_map, type_, fallback_key):
```
Comment (thibaudcolas) [inline draftjs_exporter/options.py]: Generic method so `block`, `style` and `entity` all use the same logic.

File: example.py [hunk 6]
```diff
@@ -118,6 +119,29 @@ def render(self, props):
...
+        # Directly return the block's children to keep its content.
+        return props['children']
+    elif type_ == 'example-delete':
+        logging.error('Missing config for "%s". Deleting block.' % type_)
+        # Return None to not render anything, removing the whole block.
+        return None
+    else:
+        logging.warn('Missing config for "%s". Using div instead.' % type_)
+        # Provide a fallback.
+        return DOM.create_element('div', {}, props['children'])
```
Comment (thibaudcolas) [inline example.py]: Quite a basic example but it demonstrates how you would define fallbacks depending on different scenarios, and also potentially do logging.

File: draftjs_exporter/options.py [hunk 7]
```diff
@@ -30,37 +32,38 @@ def __ne__(self, other):
...
+            config = kind_map[fallback_key]
+        else:
+            config = kind_map[type_]
 
-        if isinstance(block, list):
-            raise ConfigException('Block "%s" uses unsupported list-style config' % type_)
-        elif isinstance(block, dict):
-            if 'element' not in block:
-                raise ConfigException('Block "%s" does not define an element' % type_)
+        if isinstance(config, list):
```
Comment (loicteixeira) [inline draftjs_exporter/options.py]: Because Python encourages duck typing, this test and the following ones could fail (e.g. `isinstance(config, list)` will fail if the object is a list like but not a list, or is a `tuple`, etc), which will fallback to the `else` clause and return a mis-configured `Options` object.

Assuming that the current implementation expects an actual class (the element) when it reaches the `else` clause, I think we can flip things around like this:

```python
from collections import Mapping
from inspect import isclass

if isclass(config):
    opts = Options(type_, config)

elif isinstance(config, Mapping):
    if 'element' not in config:
        raise ConfigException('"%s" does not define an element' % type_)
    opts = Options(type_, **config)

else:
    raise ConfigException('"%s" is improperly configured, expected element or dict' % type_)
```

Comment (thibaudcolas) [inline draftjs_exporter/options.py]: This `list` check is only meant as a deprecation helper. In the interest of moving fast this week, and because I'm not sure it was even a good idea to start with, I'll get rid of it and move on.

I get the feeling your comment still holds true for the `dict` check, but let's wait for someone to shoot themselves in the foot before we fix that (the documentation is getting better, so I'm willing to think it will be good enough in this respect).

One more thing – at this stage the element can be two types of things:

- Anything that the `DOM.create_element` function accepts (tag name, function, class, class instance)
- A dict wrapping the above, with an `element` key and optionally other keys to define wrapper & props.
